### DISCUSSION TITLE
How does Iced know a widget is the same between renders?

### DISCUSSION THREAD

**Discussion by lufte** [hecrj/iced#1874] [Q&A] [answered]
With stateful widgets this wasn't a problem because the developer decided whether to assign the same state to a widget or not, but now I wonder how does Iced decide whether my focused TextInput should remain focused in the next render. I'm experiencing an issue where my TextInput is losing focus as I write into it because some other widgets appear/disappear depending on what's written, but since the widget tree changes I suspect Iced is basing this decision on the TextInput's position in the tree.

**Comment by 13r0ck** [comment, marked as answer]
Yep. Iced just assumes based by itterating through each widget's children recursively.

Give the text input widget an `.id()` and after each event created by the text input just also return a focus widget operation, that should keep the text input in focus even with layout changes
https://github.com/iced-rs/iced/blob/master/examples/todos/src/main.rs#L102

**Reply by lufte** [reply]
Thanks, I suspected that was the required solution.

**Reply by hecrj** [reply]
Eventually, the `id` should be enough for `iced` to ensure continuity; so manually refocusing will be unnecessary.

Identifiers are currently not considered during diffing, however.

**Reply by lufte** [reply]
> Eventually, the `id` should be enough for `iced` to ensure continuity; so manually refocusing will be unnecessary.
> 
> Identifiers are currently not considered during diffing, however.

Ah, in this case I'll leave a nice #TODO next to it so I remember to take it off when that feature lands :)
